File: frontend/lib/api/wiki.ts
import type { WikiGraphEdge, WikiGraphNode, WikiGraphResponse, WikiIngestResult, WikiPage, WikiSource, WikiSpace } from "@/lib/types";
import { apiFetch } from "./client";

export type WikiSpaceResponse = {
  id: string; name: string; description: string; category: WikiSpace["category"];
  page_count: number; source_count: number; chunk_count: number; updated_at: string; tags: string[];
};

export type WikiPageResponse = {
  id: string; space_id: string; title: string; type: WikiPage["type"];
  source: string; status: WikiPage["status"]; chunk_count: number; relation_count: number; updated_at: string;
};

export type WikiGraphEdgeResponse = Omit<WikiGraphEdge, "relationType" | "signals"> & {
  relation_type: string;
  signals: { direct_link: number; source_overlap: number; common_neighbor: number; type_affinity: number };
};

export type WikiGraphNodeResponse = Omit<WikiGraphNode, "linkCount"> & { link_count: number };

export type WikiGraphResponseWire = {
  nodes: WikiGraphNodeResponse[]; edges: WikiGraphEdgeResponse[];
  stats: { node_count: number; edge_count: number; page_count: number; source_count: number };
};

type WikiSourceResponse = {
  id: string; space_id: string; filename: string; status: string;
  error_message: string; page_count: number; created_at: string; updated_at: string;
};

type WikiIngestResultResponse = {
  source_id: string; status: string; page_count: number; written_paths: string[]; error?: string | null;
};

function formatDate(value: string): string {
  return new Intl.DateTimeFormat("zh-CN", { month: "short", day: "numeric", hour: "2-digit", minute: "2-digit" }).format(new Date(value));
}

function toWikiSpace(space: WikiSpaceResponse): WikiSpace { return { id: space.id, name: space.name, description: space.description, category: space.category, pageCount: space.page_count, sourceCount: space.source_count, chunkCount: space.chunk_count, updatedAt: formatDate(space.updated_at), tags: space.tags }; }
function toWikiPage(page: WikiPageResponse): WikiPage { return { id: page.id, spaceId: page.space_id, title: page.title, source: page.source, status: page.status, type: page.type, chunkCount: page.chunk_count, relationCount: page.relation_count, updatedAt: formatDate(page.updated_at) }; }
function toWikiGraph(graph: WikiGraphResponseWire): WikiGraphResponse { return { nodes: graph.nodes.map((node) => ({ id: node.id, title: node.title, type: node.type, status: node.status, source: node.source, linkCount: node.link_count })), edges: graph.edges.map((edge) => ({ source: edge.source, target: edge.target, weight: edge.weight, relationType: edge.relation_type, signals: { directLink: edge.signals.direct_link, sourceOverlap: edge.signals.source_overlap, commonNeighbor: edge.signals.common_neighbor, typeAffinity: edge.signals.type_affinity } })), stats: { nodeCount: graph.stats.node_count, edgeCount: graph.stats.edge_count, pageCount: graph.stats.page_count, sourceCount: graph.stats.source_count } }; }

export { toWikiSpace, toWikiPage, toWikiGraph };

function toSource(s: WikiSourceResponse): WikiSource {
  return { id: s.id, spaceId: s.space_id, filename: s.filename, status: s.status as WikiSource["status"], errorMessage: s.error_message, pageCount: s.page_count, createdAt: s.created_at, updatedAt: s.updated_at };
}

function toResult(r: WikiIngestResultResponse): WikiIngestResult {
  return { sourceId: r.source_id, status: r.status, pageCount: r.page_count, writtenPaths: r.written_paths, error: r.error };
}

export async function listWikiSpaces(signal?: AbortSignal): Promise<WikiSpace[]> {
  return (await apiFetch<WikiSpaceResponse[]>("/api/wiki-spaces", { signal })).map(toWikiSpace);
}

export async function listWikiPages(spaceId: string, signal?: AbortSignal): Promise<WikiPage[]> {
  return (await apiFetch<WikiPageResponse[]>(`/api/wiki-spaces/${encodeURIComponent(spaceId)}/pages`, { signal })).map(toWikiPage);
}

export async function getWikiGraph(spaceId: string, signal?: AbortSignal): Promise<WikiGraphResponse> {
  return toWikiGraph(await apiFetch<WikiGraphResponseWire>(`/api/wiki-spaces/${encodeURIComponent(spaceId)}/graph`, { signal }));
}

export async function listSources(spaceId: string, signal?: AbortSignal): Promise<WikiSource[]> {
  return (await apiFetch<WikiSourceResponse[]>(`/api/wiki-spaces/${encodeURIComponent(spaceId)}/sources`, { signal })).map(toSource);
}

export async function createSource(spaceId: string, filename: string, content: string): Promise<WikiSource> {
  return toSource(await apiFetch<WikiSourceResponse>(`/api/wiki-spaces/${encodeURIComponent(spaceId)}/sources`, {
    method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify({ filename, content }),
  }));
}

export async function deleteSource(spaceId: string, sourceId: string): Promise<void> {
  await apiFetch<{ deleted: boolean }>(`/api/wiki-spaces/${encodeURIComponent(spaceId)}/sources/${encodeURIComponent(sourceId)}`, { method: "DELETE" });
}

export async function ingestSources(spaceId: string, sourceIds: string[]): Promise<WikiIngestResult[]> {
  return (await apiFetch<WikiIngestResultResponse[]>(`/api/wiki-spaces/${encodeURIComponent(spaceId)}/ingestions`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ source_ids: sourceIds }),
  })).map(toResult);
}

File: backend/app/schemas/wiki.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class WikiSpaceRead(BaseModel):
    """前端空间列表项。"""

    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    description: str
    category: str
    page_count: int
    source_count: int
    chunk_count: int
    updated_at: datetime
    tags: list[str]


class WikiPageRead(BaseModel):
    """前端页面列表项。"""

    model_config = ConfigDict(from_attributes=True)

    id: str
    space_id: str
    title: str
    type: str
    source: str
    status: str
    chunk_count: int
    relation_count: int
    updated_at: datetime


class WikiPageSearchRead(WikiPageRead):
    """搜索结果包含正文，供查询端直接读取命中页面。"""

    content: str


class WikiGraphNode(BaseModel):
    """知识图谱节点，对应一个 WikiPage。"""

    id: str
    title: str
    type: str
    status: str
    source: str
    link_count: int


class WikiGraphEdgeSignals(BaseModel):
    """边权重的四个来源信号。"""

    direct_link: float  # 直接链接数量
    source_overlap: float  # 来源重叠度
    common_neighbor: float  # 公共邻居对数占比
    type_affinity: float  # 类型亲和度


class WikiGraphEdge(BaseModel):
    """知识图谱边，表示两节点间的关联强度。"""

    source: str
    target: str
    weight: float  # 综合权重，由四个信号加权求和
    signals: WikiGraphEdgeSignals
    relation_type: str


class WikiGraphStats(BaseModel):
    """图谱统计信息。"""

    node_count: int
    edge_count: int
    page_count: int
    source_count: int


class WikiGraphResponse(BaseModel):
    """完整知识图谱响应，不含 embedding 二进制数据。"""

    nodes: list[WikiGraphNode]
    edges: list[WikiGraphEdge]
    stats: WikiGraphStats


class WikiSourceRead(BaseModel):
    """来源列表项，包含最近任务和页面关联摘要。"""

    model_config = ConfigDict(from_attributes=True)
    id: str
    space_id: str
    filename: str
    mime_type: str
    import_kind: str
    original_uri: str
    content_size: int
    version: int
    last_job_id: str | None = None
    status: str
    error_message: str
    page_count: int
    related_page_count: int = 0
    last_job: "WikiIngestJobRead | None" = None
    created_at: datetime
    updated_at: datetime


class WikiSourceCreate(BaseModel):
    """创建文本来源的请求体。"""

    filename: str
    content: str


class WikiIngestRequest(BaseModel):
    """统一摄入请求：一个来源也放进数组，便于后续扩展选项。"""

    source_ids: list[str] = Field(min_length=1)


class WikiIngestResult(BaseModel):
    source_id: str
    job_id: str | None = None
    status: str
    page_count: int
    written_paths: list[str]
    error: str | None = None


class WikiIngestJobRead(BaseModel):
    """来源摄入任务的前端展示模型。"""

    model_config = ConfigDict(from_attributes=True)

    id: str
    space_id: str
    source_id: str
    job_type: str
    status: str
    stage: str
    progress_current: int
    progress_total: int
    error_message: str
    started_at: datetime | None
    finished_at: datetime | None
    created_at: datetime
    updated_at: datetime


class WikiSourcePageRead(BaseModel):
    """来源关联页面的轻量视图。"""

    id: str
    title: str
    type: str
    relation: str
    job_id: str | None = None
    updated_at: datetime


class WikiSourceDetailRead(WikiSourceRead):
    """来源详情，包含页面和任务链路。"""

    pages: list[WikiSourcePageRead]
    jobs: list[WikiIngestJobRead]


class WikiSourceDeleteImpact(BaseModel):
    """删除来源前的影响范围预览。"""

    source_id: str
    filename: str
    related_page_count: int
    orphan_page_count: int
    pages: list[WikiSourcePageRead]


class WikiIngestJobUpdate(BaseModel):
    """任务状态更新请求；当前只允许请求取消。"""

    status: Literal["canceled"]
